src/Microsoft.OData.Core/JsonLight/ODataJsonLightOutputContext.cs

```diff
@@ -953,32 +945,73 @@ protected override void Dispose(bool disposing)
-                if (asynchronousJsonWriterFactory is DefaultJsonWriterFactory && this.concreteJsonWriter != null)
-                {
-                    return this.concreteJsonWriter;
-                }
+                throw new ODataException(Strings.ODataMessageWriter_StreamBasedJsonWriterFactory_ReturnedNull(encoding.WebName, isIeee754Compatible));
+            }
 
-                asynchronousJsonWriter = asynchronousJsonWriterFactory.CreateAsynchronousJsonWriter(textWriter, isIeee754Compatible);
-                Debug.Assert(asynchronousJsonWriter != null, "asynchronousJsonWriter != null");
+            return jsonWriter;
+        }
+
+        private void VerifyJsonWriter()
+        {
+            // Asynchronous support is not implemented in Microsoft.Spatial library.
+            // To write spatial data, we rely on the synchronous PrimitiveConverter.Instance.WriteJsonLight(object, IJsonWriter) method.
+            // When writing asynchronously we wrap this method in a Task. WriteJsonLight method takes an
+            // IJsonWriter parameter while the asynchronous writer is declared as IJsonWriterAsync.
+            // When writing asynchronously, we have to ensure that the IJsonWriter that is used
+            // for writing spatial data is the same instance as the IJsonWriterAsync used for writing
+            // everything else in order to guarantee that the writer state is correctly maintained
+            // throughout the writing process (e.g. keeping track of the writer's current scope).
+            // When an IJsonWriterAsync is provided, it must also implement IJsonWriter so that the same instance
+            // can be reused for spatial data. If we somehow end up with 2 separate instances, then we fail early with an exception.
+            // Merging IJsonWriter and IJsonWriterAsync interface in a major release will simplify this.
+
+            if (this.jsonWriter != null && this.asynchronousJsonWriter == null && this.jsonWriter is IJsonWriterAsync jsonWriterAsync)
+            {
+                this.asynchronousJsonWriter = jsonWriterAsync;
             }
 
-            this.concreteJsonWriter = asynchronousJsonWriter as JsonWriter;
-            if (this.concreteJsonWriter != null && writerSettings.ArrayPool != null)
+            if (this.asynchronousJsonWriter != null && this.jsonWriter == null)
             {
-                this.concreteJsonWriter.ArrayPool = writerSettings.ArrayPool;
+                if (this.asynchronousJsonWriter is IJsonWriter jsonWriter)
+                {
+                    this.jsonWriter = jsonWriter;
+                }
+                else
+                {
+                    throw new ODataException(Strings.ODataMessageWriter_IJsonWriterAsync_Must_Implement_IJsonWriter);
+                }
             }
 
-            return asynchronousJsonWriter;
+            if (this.asynchronousJsonWriter != null && !this.asynchronousJsonWriter.Equals(this.jsonWriter))

```

Comment:
I assume/hope this is reference equality

Reply:
That's a good point. In most cases it should be reference equality. But we can't guarantee that. If the class happens to override `Equals` then this might not work. I have switched to the more reliable `Object.ReferenceEquals()` method.